### PR TITLE
Query params

### PR DESCRIPTION
**Pull request by MarquiseRosier**
Please ensure your pull request adheres to the following guidelines:
- [x] make sure to link the related issues in this description
- [x] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues
#11 

Thanks for contributing!

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -103,9 +104,15 @@ const wrappedFetch = async (url, options = DEFAULT_FETCH_OPTIONS) => {
     }
   }
 
+  if (opts.qs) {
+    urlWithQuery = new URL(url);
+    Object.keys(opts.qs).forEach((key) => urlWithQuery.searchParams.append(key, opts.qs[key]));
```
Comment (tripodsan) [inline src/index.js]: use `Object.entries`

Comment (tripodsan) [reply src/index.js]: maybe also support arrays? eg:
```
qs: {
  name: ['foo', 'bar']
}
```

File: test/index.test.js [hunk 2]
```diff
@@ -386,4 +386,21 @@ describe('Fetch Tests', () => {
...
+    let res;
+    try {
+      const qs = {
+        helix: 'dummy',
+        foo: 'bar',
+        rumple: 'stiltskin',
+      };
+      res = await fetch('https://httpbin.org/delay/2', { qs });
+    } catch (err) {
+      assert.fail('Querystring functionality failed', err);
```
Comment (tripodsan) [inline test/index.test.js]: why this catch? of the above throws, the test will fail anyways....

Comment (MarquiseRosier) [reply test/index.test.js]: @tripodsan I believe the code doesn't throw; and always returns

Comment (MarquiseRosier) [reply test/index.test.js]: so i should definitely fix this lol

File: src/index.js (hunk 1 above)
Comment (MarquiseRosier) [reply src/index.js]: I'll make another feature request for the array querystring :) @tripodsan

File: src/index.js [hunk 3]
```diff
@@ -103,9 +104,15 @@ const wrappedFetch = async (url, options = DEFAULT_FETCH_OPTIONS) => {
     }
   }
 
+  if (opts.qs) {
+    urlWithQuery = new URL(url);
+    Object.entries(opts.qs).forEach(([k, v]) => urlWithQuery.searchParams.append(k, v));
+    urlWithQuery = urlWithQuery.toString();
```
Comment (tripodsan) [inline src/index.js]: I think using `URL.href` is better here.

**Comment by MarquiseRosier** [comment]
Scrapping this and instead using helix-fetch's built-in URL support we can do
const { fetch } = require('@adobe/helix-fetch)
const url = new URL('https://httpbin.com');
url.searchParams.set('k', 'v', 'k1', 'v1'); 
fetch(url);

File: README.md [hunk 4]
```diff
@@ -118,6 +118,21 @@ $ npm install @adobe/helix-fetch
...
+
+const { fetch } = require('@adobe/helix-fetch');
+
+const qs = {
+  helix: 'dummy',
+  foo: 'bar',
+  rumple: "stiltskin",
+};
+
+const resp = await fetch('https://httpbin.org/delay/2', {qs});
```
Comment (stefan-guggisberg) [inline README.md]: nitpicking: I would use a different url, e.g. 

```js
const resp = await fetch('https://httpbin.org/bytes/512', { qs }); 
```

File: test/index.test.js [hunk 5]
```diff
@@ -386,4 +386,16 @@ describe('Fetch Tests', () => {
...
   });
+
+  it('fetch supports querystrings', async () => {
+    const EXPECTED = 'https://httpbin.org/delay/2?helix=dummy&foo=bar&rumple=stiltskin';
+    const qs = {
+      helix: 'dummy',
+      foo: 'bar',
+      rumple: 'stiltskin',
+    };
+    const res = await fetch('https://httpbin.org/delay/2', { qs });
```
Comment (stefan-guggisberg) [inline test/index.test.js]: nitpicking again: a delayed response is not essential for this test. I'd use a different url for testing, e.g. for random bytes: 

```js
    const resp = await fetch('https://httpbin.org/bytes/512', { qs }); 
```

File: test/index.test.js [hunk 6]
```diff
@@ -386,4 +386,16 @@ describe('Fetch Tests', () => {
...
+    const result = await res.json();
+    assert.equal(result.url, EXPECTED);
+  }).timeout(3000);
```
Comment (stefan-guggisberg) [inline test/index.test.js]: the timeout is probably not required if you change the url above.

**Comment by MarquiseRosier** [comment]
@trieloff @stefan-guggisberg so i'm holding on to my guns here; or we can go the helper function route; but then I'd want to bake it into https://github.com/adobe/helixshared/blob/master/src/utils.js

**Comment by tripodsan** [comment]
> @trieloff @stefan-guggisberg so i'm holding on to my guns here; or we can go the helper function route; but then I'd want to bake it into https://github.com/adobe/helix-shared/blob/master/src/utils.js

I don't think this is a good idea to add such a simple function to helix-shared. it doesn't offer that much improvement, I think.

what about add it as helper to helix-fetch, eg:

```
const { createURL, fetch } = require('@adobe/helix-fetch');
fetch(
  createURL('https://httpbin.org', { qs: {k: "v", k1: "v1", k2: "v2"}),
  { method: 'POST', body: {}}
);
```

File: README.md (hunk 4 above)
Comment (stefan-guggisberg) [inline README.md]:
```suggestion
const resp = await fetch('https://httpbin.org/json', {qs});
```

**Comment by trieloff** [comment]
Separate function in `helix-fetch` as @tripodsan suggested would be my preference.

**Comment by stefan-guggisberg** [comment]
> Separate function in `helix-fetch` as @tripodsan suggested would be my preference.

ok, let's do that.

File: test/index.test.js [hunk 7]
```diff
@@ -441,7 +441,10 @@ describe('Fetch Tests', () => {
     assert(!resp.fromCache);
   });
 
-  it('forcing HTTP/1(.1) works', async () => {
+  // eslint-disable-next-line func-names
+  it('forcing HTTP/1(.1) works', async function () {
```
Comment (tripodsan) [inline test/index.test.js]:
```suggestion
  it('forcing HTTP/1(.1) works', async function test() {
```

File: src/index.js [hunk 8]
```diff
@@ -78,6 +78,12 @@ function createPushHandler(ctx) {
   };
 }
 
+function createUrl(url, qs = {}) {
```
Comment (tripodsan) [inline src/index.js]: maybe add some js doc?

File: src/index.js [hunk 9]
```diff
@@ -183,6 +189,14 @@ class FetchContext {
        * Error thrown when a request timed out.
        */
       TimeoutError,
+
+      /**
+       * Create a URL with query parameters
+       *
+       * @param {string} url request url
+       * @param {object} qs request query parameters
```
Comment (tripodsan) [inline src/index.js]:
```suggestion
       * @param {object} [qs={}] request query parameters
```

File: test/index.test.js [hunk 10]
```diff
@@ -421,6 +421,24 @@ describe('Fetch Tests', () => {
...
   });
 
+  it('fetch supports querystrings', async () => {
+    const EXPECTED = 'https://httpbin.org/json?helix=dummy&foo=bar&rumple=stiltskin';
+    const qs = {
+      helix: 'dummy',
+      foo: 'bar',
+      rumple: 'stiltskin',
+    };
+    const res = await fetch(createUrl('https://httpbin.org/json', qs));
```
Comment (tripodsan) [inline test/index.test.js]: you don't need to test the `fetch` method, only the `createIUrl` method.

File: test/index.test.js [hunk 11]
```diff
@@ -421,6 +421,24 @@ describe('Fetch Tests', () => {
...
+    assert.equal(res.url, EXPECTED);
+  });
+
+  it('createUrl works without qs object', async () => {
+    const EXPECTED = 'https://httpbin.org/json';
+
+    const res = await fetch(createUrl('https://httpbin.org/json'));
```
Comment (tripodsan) [inline test/index.test.js]: you don't need to test the `fetch` method, only the `createIUrl` method.

**Pull request merged**